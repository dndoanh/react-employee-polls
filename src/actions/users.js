export const RECEIVE_USERS = "RECEIVE_USERS";
export const ADD_USER_QUESTION = "ADD_USER_QUESTION";
export const ADD_USER_ANSWER = "ADD_USER_ANSWER";

const addUserQuestion = ({ author, id }) => {
  return {
    type: ADD_USER_QUESTION,
    author,
    id,
  };
};

const addUserAnswer = (userId, questionId, answer) => {
  return {
    type: ADD_USER_ANSWER,
    userId,
    questionId,
    answer,
  };
};

const receiveUsers = (users) => {
  return {
    type: RECEIVE_USERS,
    users,
  };
};

export { addUserQuestion, addUserAnswer, receiveUsers };
