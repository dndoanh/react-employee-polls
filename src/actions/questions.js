import { saveQuestion, saveQuestionAnswer } from "../utils/api";
import { addUserAnswer, addUserQuestion } from "./users";

export const ADD_QUESTION = "ADD_QUESTION";
export const ADD_ANSWER_QUESTION = "ADD_ANSWER_QUESTION";
export const RECEIVE_QUESTIONS = "RECEIVE_QUESTIONS";

const addQuestion = (question) => {
  return {
    type: ADD_QUESTION,
    question,
  };
};

const addAnswerQuestion = (userId, qid, answer) => {
  return {
    type: ADD_ANSWER_QUESTION,
    userId,
    qid,
    answer,
  };
};

const receiveQuestions = (questions) => {
  return {
    type: RECEIVE_QUESTIONS,
    questions,
  };
};

const handleAddQuestion = (firstOption, secondOption) => {
  return async (dispatch, getState) => {
    const { authedUser } = getState();
    const question = {
      optionOneText: firstOption,
      optionTwoText: secondOption,
      author: authedUser,
    };
    const savedQuestion = await saveQuestion(question);
    dispatch(addQuestion(savedQuestion));
    dispatch(addUserQuestion(savedQuestion));
  };
};

const handleAddAnswer = (qid, answer) => {
  return async (dispatch, getState) => {
    const { authedUser } = getState();
    const answerObject = {
      authedUser: authedUser.id,
      qid: qid,
      answer,
    };
    const isAdded = await saveQuestionAnswer(answerObject);
    if (isAdded) {
      dispatch(addAnswerQuestion(authedUser.id, qid, answer));
      dispatch(addUserAnswer(authedUser.id, qid, answer));
    }
  };
};

export { handleAddQuestion, handleAddAnswer, receiveQuestions };
