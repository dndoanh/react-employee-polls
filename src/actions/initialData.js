import { getInitialData } from "../utils/api";
import { receiveUsers } from "./users";
import { receiveQuestions } from "./questions";

const receiveInitialData = () => {
  return async (dispatch) => {
    const { users, questions } = await getInitialData();
    dispatch(receiveUsers(users));
    dispatch(receiveQuestions(questions));
  };
}

export { receiveInitialData };
