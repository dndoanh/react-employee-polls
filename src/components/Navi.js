import { connect } from "react-redux";
import { handleLogoutAuthedUser } from "../actions/authedUser";
import { Button, Container, Nav, Navbar } from "react-bootstrap";
import { Link } from "react-router-dom";

const Navi = ({ dispatch, authedUser }) => {
  const handleLogout = () => {
    dispatch(handleLogoutAuthedUser());
  };

  return (
    <>
      {authedUser && (
        <Navbar>
          <Container>
            <Link to={"/"} className="navbar-brand">Employee Poll</Link>
            <Nav className="me-auto">
              <Link to={"/"} className="nav-link">Home</Link>
              <Link to={"/leaderboard"} className="nav-link">Leaderboard</Link>
              <Link to={"/add"} className="nav-link">New Poll</Link>
            </Nav>
            <Navbar.Collapse className="justify-content-end">
              <Navbar.Text>
                Signed in as: {authedUser?.name}
                <Button onClick={handleLogout} variant="link">
                  Logout
                </Button>
              </Navbar.Text>
            </Navbar.Collapse>
          </Container>
        </Navbar>
      )}
    </>
  );
};

const mapStateToProps = ({ authedUser }) => ({
  authedUser,
});

export default connect(mapStateToProps)(Navi);
