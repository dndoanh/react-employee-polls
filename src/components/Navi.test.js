import { render, screen } from "@testing-library/react";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import React from "react";
import Navi from "./Navi";
import { createStore } from "redux";
import middleware from "../middlewares";
import reducer from "../reducers";
import { loginAuthedUser } from "../actions/authedUser";

const store = createStore(reducer, middleware);

describe("Navi", () => {
  test("should render the components", () => {
    store.dispatch(
      loginAuthedUser({
        id: "tylermcginnis",
        name: "Tyler Alibaba",
      })
    );
    const view = render(
      <Provider store={store}>
        <BrowserRouter>
          <Navi />
        </BrowserRouter>
      </Provider>
    );
    expect(view).toBeDefined();
    expect(view).toMatchSnapshot();
  });

  test("should display all elements", () => {
    store.dispatch(
      loginAuthedUser({
        id: "tylermcginnis",
        name: "Tyler Alibaba",
      })
    );
    render(
      <Provider store={store}>
        <BrowserRouter>
          <Navi />
        </BrowserRouter>
      </Provider>
    );

    const homeLinkElement = screen.getByTestId("home-link");
    const newPollLinkElement = screen.getByTestId("new-poll-link");
    const leaderboardLinkElement = screen.getByTestId("leaderboard-link");
    const logoutLinkElement = screen.getByTestId("logout-link");

    expect(homeLinkElement.textContent).toBe("Home");
    expect(newPollLinkElement.textContent).toBe("New Poll");
    expect(leaderboardLinkElement.textContent).toBe("Leaderboard");
    expect(logoutLinkElement.textContent).toBe("Logout");
  });
});
