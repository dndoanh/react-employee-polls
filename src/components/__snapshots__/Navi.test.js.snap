// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navi should render the components 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <nav
        class="navbar navbar-expand navbar-light"
      >
        <div
          class="container"
        >
          <a
            class="navbar-brand"
            href="/"
          >
            Employee Poll
          </a>
          <div
            class="me-auto navbar-nav"
          >
            <a
              class="nav-link"
              data-testid="home-link"
              href="/"
            >
              Home
            </a>
            <a
              class="nav-link"
              data-testid="leaderboard-link"
              href="/leaderboard"
            >
              Leaderboard
            </a>
            <a
              class="nav-link"
              data-testid="new-poll-link"
              href="/add"
            >
              New Poll
            </a>
          </div>
          <div
            class="justify-content-end navbar-collapse collapse"
          >
            <span
              class="navbar-text"
            >
              Signed in as: 
              Tyler Alibaba
              <button
                class="btn btn-link"
                data-testid="logout-link"
                type="button"
              >
                Logout
              </button>
            </span>
          </div>
        </div>
      </nav>
    </div>
  </body>,
  "container": <div>
    <nav
      class="navbar navbar-expand navbar-light"
    >
      <div
        class="container"
      >
        <a
          class="navbar-brand"
          href="/"
        >
          Employee Poll
        </a>
        <div
          class="me-auto navbar-nav"
        >
          <a
            class="nav-link"
            data-testid="home-link"
            href="/"
          >
            Home
          </a>
          <a
            class="nav-link"
            data-testid="leaderboard-link"
            href="/leaderboard"
          >
            Leaderboard
          </a>
          <a
            class="nav-link"
            data-testid="new-poll-link"
            href="/add"
          >
            New Poll
          </a>
        </div>
        <div
          class="justify-content-end navbar-collapse collapse"
        >
          <span
            class="navbar-text"
          >
            Signed in as: 
            Tyler Alibaba
            <button
              class="btn btn-link"
              data-testid="logout-link"
              type="button"
            >
              Logout
            </button>
          </span>
        </div>
      </div>
    </nav>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
